stdlib/sys.pyi

```diff
@@ -27,12 +28,11 @@ _T = TypeVar("_T")
 # The following type alias are stub-only and do not exist during runtime
 _ExcInfo = Tuple[Type[BaseException], BaseException, TracebackType]
 _OptExcInfo = Union[_ExcInfo, Tuple[None, None, None]]
-_PathSequence = Sequence[Union[bytes, str]]
+_ModulePath = MutableSequence[str]  # Same as `types.ModuleType.__path__`
 
 # Unlike importlib.abc.MetaPathFinder, invalidate_caches() might not exist (see python docs)
 class _MetaPathFinder(Protocol):
-    def find_module(self, fullname: str, path: _PathSequence | None) -> Loader | None: ...
-    def find_spec(self, fullname: str, path: _PathSequence | None, target: ModuleType | None = ...) -> ModuleSpec | None: ...
+    def find_spec(self, fullname: str, path: _ModulePath | None, target: ModuleType | None = ...) -> ModuleSpec | None: ...
```

Comment:
This doesn't seem correct to me. A MetaPathFinder must also have the older `find_module` method, and as the docs say, `importlib.abc.MetaPathFinder` provides it if only `find_spec` is defined.

Reply:
Agreed - the docs imply that `importlib.abc.MetaPathFinder` provides a default implementation for `find_module`, but there's no reason an entry on `sys.meta_path` has to inherit from that ABC.

Also, this still presumes the API is a `find_spec` *method*. That's not necessarily true - you can put the class itself on `sys.meta_path` and then supply a `classmethod` or a `staticmethod`.

Reply:
That's not something we can fix in typeshed though: type checkers will have to support passing a class that matches a protocol. That's python/mypy#4536.

Reply:
`importlib.abc.MetaPathFinder` provides it for backward compatibility with code for Python < 3.5. Meta path finders are not required to implement `find_module`.

Reply:
That seems reasonable, but where is that documented? The docs say "Deprecated since 3.4" but there's no documentation I can see that says the following code is illegal (as opposed to deprecated):

```
for hook in sys.meta_path:
    hook.find_module(fullname, path)
```

The problem is that there's a difference between how the standard import machinery uses meta path finders (it never calls `find_module`) and what user-written code is allowed to do with them (they can do whatever they like, as long as they follow the documentation).

I don't really mind either way, but I feel that having typeshed implement a rule that doesn't match the documented API could be the cause of bugs in the future.

Reply:
> type checkers will have to support passing a class that matches a protocol. That's python/mypy#4536.

It seems to happen with pyright, too.

Reply:
I'm not sure why you think it's illegal? Protocols are not exhaustive - you can implement any other method of your liking on the meta path finder class.

Reply:
Mostly because I don't know much about typing 😉 And a little bit because I wasn't thinking straight 😉

Yes, if the idea is that the type defines the minimum that *must* be present on the hook entry, it seems fair to omit `find_module`.

Reply:
The issue here is that protocols do not have a concept of optional members; we have to choose between specifying `find_module` on the protocol, meaning that it has to be implemented, or for it not to be specified and for `sys.meta_path` entries to potentially lie about _not_ having `find_module`.

Reply:
>It seems to happen with pyright, too.

If you need this fixed, it might be good to report it on pyright's github. Pyright issues usually get fixed *and released* in a couple days or so.

>there's no documentation I can see that says the following code is illegal (as opposed to deprecated):

I agree. Unless we have a practical reason to delete `find_module` (i.e. some real-world non-toy-example code doesn't type check with it), we should keep it.

Reply:
Well, my code (https://github.com/pfmoore/editables) has a metapath finder that doesn't define `find_module`. That's how I found this. I don't consider my code wrong, although I *also* don't mind inheriting from `MetaPathFinder`, so I'm not hugely bothered either way.

Regarding classmethods,

> If you need this fixed, it might be good to report it on pyright's github.

I don't need it fixed as such. I don't use pyright myself (I use mypy for any static checks I do - I only tried pyright as a one-off check). Someone who does use pyright can probably raise the issue with them. However, I actually think it should be made clear in the `Protocol` documentation (if it isn't already) that this usage is valid - but I don't know where to find the protocol documentation, or how to raise an issue with it...

Reply:
Now it makes sense to me.